l1-contracts/test/merkle/helpers/MerkleLibHelper.sol

```diff
@@ -55,7 +56,63 @@ contract MerkleLibHelper {
     return (min, max);
   }
 
+  /**
+   * @notice Computes the root for a binary unbalanced Merkle-tree given the leaves.
+   * @dev Filled in greedily with subtrees. Useful for outHash tree.
+   * @param _leaves - The 32 bytes leafs to build the tree of.
+   * @return The root of the Merkle tree.
+   */
   function computeUnbalancedRoot(bytes32[] memory _leaves) external pure returns (bytes32) {
-    return MerkleLib.computeUnbalancedRoot(_leaves);
+    // e.g. an unbalanced tree of 7 txs will contain subtrees of 4, 2, and 1 tx(s) = 111
+    // e.g. an unbalanced tree of 9 txs will contain subtrees of 8 and 1 tx(s) = 1001
+    // We collect the roots of each subtree
+    bytes32 root;
+    uint256 currentSubtreeSize = 1;
+    uint256 numTxs = _leaves.length;
+    // We must calculate the smaller rightmost subtrees first, hence starting at 1
+    while (numTxs != 0) {
+      // If size & txs == 0, the subtree doesn't exist for this number of txs
+      if (currentSubtreeSize & numTxs == 0) {
+        currentSubtreeSize <<= 1;
+        continue;
+      }
+      bytes32[] memory leavesInSubtree = new bytes32[](currentSubtreeSize);
+      uint256 start = numTxs - currentSubtreeSize;
+      for (uint256 i = start; i < numTxs; i++) {
+        leavesInSubtree[i - start] = _leaves[i];
+      }
+      bytes32 subtreeRoot = computeRoot(leavesInSubtree);
+      root = numTxs == _leaves.length ? subtreeRoot : Hash.sha256ToField(bytes.concat(subtreeRoot, root));
+      numTxs -= currentSubtreeSize;
+      currentSubtreeSize <<= 1;
+    }
+    return root;
+  }
+
+  /**
+   * @notice Computes the root for a binary Merkle-tree given the leafs.
+   * @dev Uses sha256.
+   * @param _leafs - The 32 bytes leafs to build the tree of.
+   * @return The root of the Merkle tree.
+   */
+  function computeRoot(bytes32[] memory _leafs) internal pure returns (bytes32) {
+    // @todo Must pad the tree
+    uint256 treeDepth = 0;
+    while (2 ** treeDepth < _leafs.length) {
+      treeDepth++;
+    }
+    uint256 treeSize = 2 ** treeDepth;
+    assembly {
+      mstore(_leafs, treeSize)
+    }
+
+    for (uint256 i = 0; i < treeDepth; i++) {
+      for (uint256 j = 0; j < treeSize; j += 2) {
+        _leafs[j / 2] = Hash.sha256ToField(bytes.concat(_leafs[j], _leafs[j + 1]));
+      }
+      treeSize /= 2;
+    }
+
+    return _leafs[0];
```

Comment:
I guess this was already covered, but isn't there a risk of reading dirty memory by forcefully resizing the array, if there was something written _after_ the leafs array?

Reply:
There is. But it is a test only contract you are reading.